Update Admission Control docs for RACv2 (#19368)
cockroachdb/docs: src/current/v25.1/admission-control.md

- [Bulk data imports]({% link {{ page.version.version }}/import-into.md %}).
- [`COPY`]({% link {{ page.version.version }}/copy.md %}) statements.
- [Deletes]({% link {{ page.version.version }}/delete-data.md %}) (including deletes initiated by [row-level TTL jobs]({% link {{ page.version.version }}/row-level-ttl.md %}); the [selection queries]({% link {{ page.version.version }}/selection-queries.md %}) performed by TTL jobs are also subject to CPU admission control).
- [Backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}).
- [Schema changes]({% link {{ page.version.version }}/online-schema-changes.md %}), including index and column backfills (on both the [leaseholder replica]({% link {{ page.version.version }}/architecture/replication-layer.md %}#leases) and [follower replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)).
- [Follower replication work](#replication-admission-control).

> - [Raft log entries being written to disk]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft).

Author: @kvoli is this 'raft log entries' sort of subsumed by 'follower replication work'? or does it make sense to continue to leave it alone as a separate item?

Reviewer: `follower replication work` includes `Raft log entries being written to disk` but the leader also paces writing log entries to disk.

Having said that, I'd lean towards leaving both despite the overlap because it informs the reader to think about both.

Author: 👍 - thanks, will leave both in that case

- [Changefeeds]({% link {{ page.version.version }}/create-and-configure-changefeeds.md %}).
- [Intent resolution]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#write-intents).

{{site.data.alerts.callout_info}}
Admission control is beneficial when overall cluster health is good but some nodes are experiencing overload. If you see these overload scenarios on many nodes in the cluster, that typically means the cluster needs more resources.
{{site.data.alerts.end}}

> ### Replication admission control

Author: @kvoli and @nhamlani101 : i added this new section as we discussed but I'm not sure it's 100% aligned with what we talked about, happy to update with any feedback.  i wanted to provide the user more detail to 1) head off questions and 2) make clear what the changes/benefits of coming up to v25.1 are (idk if this stuff is being backported or not, AIUI it's v25.1 only for now?)

Reviewer: >  [...] 2) make clear what the changes/benefits of coming up to v25.1 are (idk if this stuff is being backported or not, AIUI it's v25.1 only for now?)

Its only 25.1 for the user noticeable behavior change but we did apply a migration w/o a behavior change in 24.3 to the new system (added context for you but don't think its important to mention).

Author: thanks for explaining that. i will leave it as is per your guidance here then

The admission control subsystem paces all work done at the [Replication Layer]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) to avoid cluster overload. This includes user-facing writes from SQL statements, as well as background (elastic) replication work.

> {% include_cached new-in.html version="v25.1" %} This includes controlling the pacing of catchup writes at the replication layer to avoid overloading slow or [newly restarted nodes]({% link {{ page.version.version }}/eventlog.md %}#node_restart) with replication flows. Note that this pacing does not slow down user-facing SQL writes; it only ensures there are fewer impacts from background operations.

Reviewer: ```
Note that this pacing does not slow down user-facing SQL writes; it only ensures there are fewer impacts from background operations.
```

Admission control will slow down user facing SQL writes in the case where all quorum required replicas are on overloaded stores. It won't result in a slow down in aggregate over time, compared to just running without admission control. So the statement is factual (imo) but may lead to mismatched expectations. e.g.,

```
# replication factor is 3
# leader is on store 1 (s1)
s1=slow* s2=slow* s3=not slow

will pace writes

# * could be an inverted LSM
```

All that being said, its probably reasonable to leave as is. wdyt @sumeerbhola?

> {% include_cached new-in.html version="v25.1" %} This includes controlling the pacing of catchup writes at the replication layer to avoid overloading slow or [newly restarted nodes]({% link {{ page.version.version }}/eventlog.md %}#node_restart) with replication flows. Note that this pacing does not slow down user-facing SQL writes; it only ensures there are fewer impacts from background operations.

Reviewer 2: Suggestion, as it was slightly hard to parse. Also to avoid having two consecutive "This includes..." sentences.

```suggestion
{% include_cached new-in.html version="v25.1" %} The pacing of catchup writes is controlled at the replication layer to avoid overloading slow or [newly restarted nodes]({% link {{ page.version.version }}/eventlog.md %}#node_restart) with replication flows. Note that this pacing does not slow down user-facing SQL writes; it only ensures there are fewer impacts from background operations.
```

Author: makes sense, applied, thanks!

At a high level, replication admission control works by:

- Pacing regular SQL writes at the rate of replica quorum. (**New in v25.1**)
- Pacing background (elastic) replication at the rate of the slowest replica.

This has the following effects:

1. Does not overload slow/restarted nodes with replication flows. (**New in v25.1**)
2. Isolates performance between regular and elastic traffic.
3. Paces regular writes at quorum speed. (**New in v25.1**)
4. Paces elastic writes at the slowest replica's speed.

> For example, prior to CockroachDB v25.1, when a leader and follower replica were disconnected from each other due to a node going away and coming back, when the follower came back the leader would slam the follower with any Raft entries it had missed. In v25.1 and later, the leader paces the entries it sends to the follower. This has the result that baseline cluster QPS (queries per second) and latency should not change substantially during perturbations such as [node restarts]({% link {{ page.version.version }}/eventlog.md %}#node_restart).

Reviewer 2: ```suggestion
For example, prior to CockroachDB v25.1, when a leader and follower replica were disconnected from each other due to a node going away and coming back, once the follower came back the leader would slam the follower with any Raft entries it had missed. In v25.1 and later, the leader paces the entries it sends to the follower. The result is that baseline cluster QPS (queries per second) and latency should not change substantially during perturbations such as [node restarts]({% link {{ page.version.version }}/eventlog.md %}#node_restart).
```

Author: updated!

> To monitor the behavior of replication admission control, see [UI Overload Dashboard &gt; Replication Admission Control]({% link {{ page.version.version }}/ui-overload-dashboard.md %}#admission-queueing-delay-p99-replication-admission-control).

Reviewer 2: ```suggestion
To monitor the behavior of replication admission control, refer to [UI Overload Dashboard &gt; Replication Admission Control]({% link {{ page.version.version }}/ui-overload-dashboard.md %}#admission-queueing-delay-p99-replication-admission-control).
```

Author: ✅
